### DISCUSSION TITLE
Ability to set trailing slash config per-page

### DISCUSSION THREAD

**Discussion by myylow** [vercel/next.js#23971] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

We have a large site powered by Next.js (thanks!), with quite a complex legacy URL structure. Some URLs are required to have trailing slashes and some are required not to, depending on the type of URL it is. This currently doesn't seem possible to achieve with Next.js alone. We achieve it by running Express in front of Next, but we'd like to remove that if possible. We also run into redirect loops sometimes, with our Express setup adding a trailing slash and then Next removing it. 

## Describe the solution you'd like

The ideal solution for me would be a way to set the trailing slash config to be customisable per-page, either in each page file itself or in next.config.js.
A less-ideal solution would just be a way to opt-out of Next.js trailing slash behavior entirely. This way, we could at least manage the URLs from a custom server without hitting a redirect loop. 

## Describe alternatives you've considered

We could change all our URLs to be fully consistent, but we are concerned about the SEO hit. I'm not aware of any other alternative we can try.

**Comment by lokst** [comment]
I am interested in this feature too. This issue looks similar to https://github.com/vercel/next.js/issues/15148

**Comment by atomicleopard** [comment]
We have the same need - trying to maintain legacy url structure which has or does not have trailing slashes on a per-path basis.

I had naively thought that creating the relevant pages structure would give me the expected behaviour - e.g. `/path/[slug].tsx` would be distinct from `/path/[slug]/index.tsx`. 

The current behaviour is a bit surprising to me, as I would expect NextJS to respect the sibling/parent relationship of my pages

e.g. In my mind `/path/[slug].tsx` should be a parent of both `/path/[slug]/index.tsx` and `/path/[slug]/foo.tsx`. 

In particular I would expect the nature of a relative link in index.tsx to be consistent with normal folder/page structures - if I had a `<a href='foo'>Link</a>` in index, I would expect to go to `/path/[slug]/foo` not `/path/foo`

The default behaviour of Next, which is to normalise one way or the other is a bit frustrating - if it did nothing I could probably work around it. If there were a setting which made Next less opinionated it would make upgrading from older platforms easier - i.e. #15148

**Comment by danidelalin** [comment]
We also need this feature before we can migrate to Next.js 
Is it on your roadmap?
